File: algorithms.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long

/*
Fast IO: 
ios::sync_with_stdio(0), cin.tie(0);
*/

/*
Properties of mod:
(a+b) % m = (a mod m + b mod m) mod m
(a*b) % m = (a mod m * b mod m) mod m
*/

/*
Data Structures
vector - random access
deque - lrandom access, ike a vector but includes push_front and pop_front methods
list - Linked list, data is not contiguous. Fast insertion and deletion, no random access.
stack/queue - provide an interface that wraps around deque
bitset - optimized array of booleans
map/set
multiset - set with duplicates
priority_queue

adjacency matrix - Can only be used if V is known. V*V space complexity, only feasible if V < 1000. O(V) to enumerate neighbors
*/

/*
STL Methods
Sorting: sort, partial_sort (sort only a range), stable_sort (maintain initial order of items that are equal)
Searching: upper_bound (iterator pointing to first element greater than val in log n complexity, must be sorted beforehand), 
lower_bound (opposite from upper_bound), binary_search
*/

/*
Debug a segfault
g++ -g program.cpp
gdb ./a.out
run
backtrace
*/

vector<ll> sieve_erastothenes(unsigned ll limit){
    vector<bool> prime(limit, true);
    for (int p = 2; p*p <= limit; p++){
        if (prime[p]){
            for (int i = p*p; i <= limit; i += p){
                prime[i] = false;
            }
        }
    }

}

int num_divisors(ll n){
    vector<ll> primes = sieve_erastothenes(n);
    int total = 1;
    for (ll p : primes){
        int count = 0;
        while (n % p == 0){
            n = n / p;
            count++;
        }
        total *= (count + 1);
    }

    return total;
}


bool is_prime(ll n) { 
    if (n <= 1)  return false; 
    if (n <= 3)  return true; 
    if (n % 2 == 0 || n % 3 == 0) return false; 
  
    for (ll i = 5; i * i <= n; i += 6) 
        if (n % i == 0 || n % (i+2) == 0) 
           return false; 
  
    return true; 
} 

class UnionFind {
    
    public:
        vector<int> id, size;
        int components;
        UnionFind(int n){
            for (int i = 0; i < n; i++){
                id.push_back(i);
                size.push_back(1);
            }
            components = n;
        }

        int find(int a){
            int root = a;
            while (id[root] != root) {
                root = id[root];
            }

            //Path compression
            while (a != root){
                int next = id[a];
                id[a] = root;
                a = next;
            }

            return root;
        }

        void unite(int a, int b){
            int root_a = find(a), root_b = find(b);
            if (root_a == root_b){
                return;
            }

            if (size[root_a] > size[root_b]){
                id[root_b] = root_a;
                size[root_a] += size[root_b];
            } else {
                id[root_a] = root_b;
                size[root_b] += size[root_a];
            }

            components--;
        }

        bool connected(int a, int b){ return find(a) == find(b);}

        int getSize(int a){return size[find(a)];}
};

bool power_of_two(ll x){
    return ((x & (x - 1)) == 0);
}

void rotate(vector<vector<int>> &vec){
    for (int i = 0; i < vec.size(); i++){
        for (int j = i + 1; j < vec.size(); j++){
            swap(vec[i][j], vec[j][i]);
        }
    }

    for (int r = 0; r < vec.size() / 2; r++){
        swap(vec[r], vec[vec.size() - r - 1]);
    }
}


vector<vector<int>> all_subsets(vector<int> vec){
    vector<vector<int>> subsets;
    int n = vec.size();
    // 1 << n is the same as 2**n
    for (int i = 0; i < (1 << n); i++){
        subsets.push_back(vector<int>());
        for (int j = 0; j < n; j++){
            if (i & (1 << j)){ //check if bit j is set
                subsets.back().push_back(vec[j]);
            }
        }
    }
    return subsets;
}

// class SegmentTree {
//     //Sum implementation with lazy propagation

//     public:
//         vector<int> arr, tree, lazy;
//         SegmentTree(vector<int> arr){
//             this->arr = arr;
//             tree.assign(arr.size()*4, -1);
//             lazy.assign(arr.size()*4, 0);
//             build(0, arr.size()-1);
//         }

//         void build(int l, int r, int node=1){
//             if (l == r){
//                 tree[node] = arr[l];
//             } else {
//                 int mid = (l + r) / 2;
//                 build(l, mid, node*2);
//                 build(mid + 1, r, node*2+1);
//                 tree[node] = tree[node*2] + tree[node*2+1]; //For sum queries
//             }
//         }

//         int rsq(int i, int j){
//             return rsq(i, j, 0, arr.size()-1);
//         }

//         int rsq(int i, int j, int l, int r, int node=1){
//             //i, j is the range for the query. l, r are the bounds for the current node
//             //i and j are both inclusive
//             update_and_propagate(l, r, node);

//             if (i <= l && j >= r){ //current interval is contained fully in [i..j]
//                 return tree[node];
//             } else if (i > r || j < l) { //current interval is completely out of [i..j]
//                 return 0; //0 because it shouldn't affect the sum query. If doing a max query, return -infinity.
//             }

//             //current interval is contained partially in [i..j]
//             int mid = (l + r) / 2;
//             int left = rsq(i, j, l, mid, node*2);
//             int right = rsq(i, j, mid + 1, r, node*2+1);
//             return left + right; //combine according to query type
//         }

//         void add_range(int i, int j, int val){
//             add_range(i, j, 0, arr.size()-1, val);
//         }

//         void add_range(int i, int j, int l, int r, int val, int node=1){
//             update_and_propagate(l, r, node);

//             if (i <= l && j >= r){ //current interval is contained fully in [i..j]
//                 tree[node] += (r - l + 1) * val;//do update
//                 if (l != r){//if we're not at a leaf
//                     lazy[node*2] += val;//mark update to be done later in both child nodes
//                     lazy[node*2+1] += val;
//                 }
//                 return;
//             } else if (i > r || j < l) { //current interval is completely out of [i..j]
//                 return; 
//             }

//             int mid = (l + r) / 2;
//             add_range(i, j, l, mid, val, node*2);
//             add_range(i, j, mid+1, r, val, node*2+1);
//             tree[node] = tree[node*2] + tree[node*2+1];
//         }

//         //Checks if an update is pending for the current node and does it if necessary. Then it propagates the update downwards
//         void update_and_propagate(int l, int r, int node){
//             if (lazy[node] != 0){//If pending updates to current node
//                 tree[node] += (r - l + 1) * lazy[node];//do the update.
//                 if (l != r){//not a leaf node
//                     lazy[node*2] += lazy[node];//propagate the update one level downwards
//                     lazy[node*2+1] += lazy[node];  
//                 }
//                 lazy[node] = 0;//remove the update for this node 
//             }
//         }
// };

class SegmentTree {

    public:
        vector<int> arr, tree;
        SegmentTree(vector<int> arr){
            this->arr = arr;
            tree.assign(arr.size()*4, -1);
            build(0, arr.size()-1);
        }

        void build(int l, int r, int node=1){
            if (l == r){
                tree[node] = arr[l];
            } else {
                int mid = (l + r) / 2;
                build(l, mid, node*2);
                build(mid + 1, r, node*2+1);
                tree[node] = max(tree[node*2], tree[node*2+1]); //For sum queries
            }
        }

        void update(int i, int v){
            update(i, 0, arr.size()-1, v);
        }

        void update(int i, int l, int r, int val, int node=1){
            if (l == r){
                tree[node] = val;
            } else {
                int mid = (l + r) / 2;
                if (i <= mid){
                    update(i, l, mid, val, node*2);
                } else {
                    update(i, mid+1, r, val, node*2+1);
                }

                tree[node] = tree[node*2] + tree[node*2+1];
            }
        }
};

void bfs(vector<vector<int>> graph, int start){
    vector<bool> visited(graph.size(), false);
    queue<int> search;
    search.push(start);

    while (!search.empty()){
        int node = search.front();
        search.pop();
        if (!visited[node]){
            visited[node] = true;
            for (int neighbor : graph[node]){
                search.push(neighbor);
            }
        }
    }
}



int main(){
    vector<int> arr = {2, 2, 4, 4, 4, 4, 1, 3, 3, 3};
    SegmentTree tree(arr);
    for (auto x : tree.tree) cout << x << " ";
    cout << endl;
}

File: algorithms.py
from timeit import timeit
from collections import *
import itertools
import bisect

#bisect.bisect_left(arr, val) same as std:lower_bound. bisect_right same as std::upper_bound


def bfs(graph, start, end):
    visited = set()
    search = deque()
    search += graph[start]

    while search:
        node = search.popleft()
        if node not in visited:
            if node == end:
                return True
            visited.add(node)
            search += graph[node]

    return False

def bfs_with_path(graph, start, end):
    parent = {}
    search = deque()
    visited = set()
    search += start
    while search:
        node = search.popleft()
        if node not in visited:
            visited.add(node)
            if node == end:
                return backtrack_path(parent, start, end)
            
            for adjacent in graph[node]:
                parent[adjacent] = node
                search += adjacent
    return backtrack_path(parent, start, end)

def backtrack_path(parent, start, end):
    path = deque()
    path.append(end)
    curr = parent[end]
    while path[0] != start:
        path.appendleft(curr)
        curr = parent[curr]
    return path


graph = {
        "A": ["B", "C"],
        "B": ["A"],
        "C": ["B"]
        }

def dfs_has_path(graph, start, end, visited=None):
    if not visited:
        visited = set()
    if start == end:
        return True
    if start in visited:
        return False
    visited.add(start)

    for adjacent in graph[start]:
        if dfs_has_path(graph, adjacent, end, visited):
            return True
    return False

# https://www.nayuki.io/page/next-lexicographical-permutation-algorithm
def next_permutation(arr):
    arr = list(arr)
    i = len(arr) - 1
    while i > 0 and arr[i - 1] >= arr[i]:
        i -= 1
    if i <= 0:
        print("No permutation, it is already the biggest one")
        return False

    pivot = i - 1
    j = len(arr) - 1
    while arr[j] <= arr[pivot]:
        j -= 1

    arr[pivot], arr[j] = arr[j], arr[pivot]
    arr[i:] = arr[len(arr) - 1: pivot: -1]
    return arr

import heapq

def dijkstra(graph, start_vertex):
    distances = {vertex: float("inf") for vertex in graph}
    distances[start_vertex] = 0
    priority_queue = []

    for vertex, distance in distances.items():
        entry = (distance, vertex)
        heapq.heappush(priority_queue, entry)

    while len(priority_queue) > 0:
        weight, current_vertex = heapq.heappop(priority_queue)
        if weight == float("inf"): break

        for neighbor, weight in graph[current_vertex].items():
            if distances[current_vertex] + weight < distances[neighbor]:
                distances[neighbor] = distances[current_vertex] + weight

    return distances


def bellman_ford_negative_cycle(graph, start):
    distances = {vertex: float("inf") for vertex in graph}
    distances[start] = 0

    for i in range(len(graph) - 1):
        for vertex in graph:
            for neighbor, weight in graph[vertex].items():
                if distances[vertex] + weight < distances[neighbor]:
                    distances[neighbor] = distances[vertex] + weight

    for vertex in graph:
            for neighbor, weight in graph[vertex].items():
                if distances[vertex] + weight < distances[neighbor]:
                    return True

    return False

def longest_increasing_subsequence(arr):
    dp = [1] * len(arr)

    for i in range(1, len(arr)):
        for j in range(i):
            if arr[i] > arr[j]:
                dp[i] = max(dp[j] + 1, dp[i])

    return max(dp)

#0 Indexed
class UnionFind():

    def __init__(self, size):
        self.id = [i for i in range(size)]
        self.__size = [1 for i in range(size)]
        self.__components = size
    
    def union(self, a, b):
        root_a, root_b = self.find(a), self.find(b)
        if root_a == root_b:
            return #Already connected
        if self.__size[root_a] > self.__size[root_b]:
            self.id[root_b] = root_a
            self.__size[root_a] += self.__size[root_b]
        else:
            self.id[root_a] = root_b
            self.__size[root_b] += self.__size[root_a]

        self.__components -= 1

    def connected(self, a, b):
        return self.find(a) == self.find(b)
    
    def find(self, a):
        root = a
        while self.id[root] != root:
            root = self.id[root]

        #Path Compression
        while a != root:
            _next = self.id[a]
            self.id[a] = root
            a = _next

        return root

    def components(self):
        return self.__components

    def size(self, a):
        return self.__size[self.find(a)]


arr = [1, 3, -2, 8, -7]
tree = [0] * (len(arr) * 4)

def build(lo, hi, node=1):
    if (lo == hi):
        tree[node] = arr[lo]
        return
    
    mid = (lo + hi) // 2
    build(lo, mid, 2*node)
    build(mid + 1, hi, 2*node+1)
    tree[node] = tree[2*node] + tree[2*node+1]

def sum_query(lo, hi, start, end, node=1):
    if start > hi or end < lo:
        return 0
    if start == lo and end == hi:
        return tree[node]
    
    mid = (lo + hi) // 2
    left_sum = sum_query(lo, mid, start, min(end, mid), 2*node)
    right_sum = sum_query(mid+1, hi, max(start, mid+1), end, 2*node+1)
    return left_sum + right_sum

def update(lo, hi, idx, val, node=1):
    if lo == hi:
        tree[node] = val
        arr[idx] = val
        return
    
    mid = (lo + hi) // 2
    if idx <= mid:
        update(lo, mid, idx, val, node*2)
    else:
        update(mid + 1, hi, idx, val, node*2+1)
    
    tree[node] = tree[2*node] + tree[2*node+1]

def prime(n):
    if n <= 1: return False
    if n <= 3: return True
    if n % 2 == 0 or n % 3 == 0: return False

    i = 5
    while i ** 2 <= n:
        if n % i == 0 or n % (i+2) == 0:
            return False
        i += 6
    
    return True

def kruskal_minspantree(edges: list):
    #greedy algorithm
    #1: sort the edges of the graph by their weights
    #2: start adding edges to the mst starting from the lowest weight
    #3: if an edge creates a cycle in the mst so far, don't add it. UFDS can be 
    #used to check if a cycle is created.
    mst_graph = UnionFind()
    mst = deque()
    cost = 0
    edges.sort(key=lambda x: x[2]) #Assuming each edge is a (u, v, w) tuple
    for edge in edges:
        u, v, w = edge
        if not mst_graph.connected(u, v):
            mst.append(edge)
            cost += w
            mst_graph.union(u, v)

    #if the length of mst if 0, no mst exists. 
    #keep in mind that if the graph is not connected, this would yield two different
    #mst's. Check for connectedness before using UFDS.components() == 1
    
    return cost #or mst if the actual tree is needed

def matrix_prefix_sum(matrix: list):
    #O(n) runtime
    #dp[i, j] represents the sum of the rectangle spanning from (0, 0) to (i, j)
    #dp[i, j] = dp[i-1][j] + dp[i][j-1] + matrix[i-1][j-1] - dp[i-1][j-1]
    #sum_region(r1, c1, r2, c2) = dp[r2][c2] - dp[r2][c1-1] - dp[r1-1][c2] + dp[r1-1][c1-1]
    #remember that dp size is (n+1) x (n+1), so the indices in matrix are shifted by one. sum_query assumes
    #that the indices have already been shifted
    n = len(matrix) 
    dp = [[0 for i in range(n + 1)] for j in range(n + 1)]

    for i in range(1, n+1):
        for j in range(1,n+1):
            dp[i][j] = dp[i-1][j] + dp[i][j-1] + matrix[i-1][j-1] - dp[i-1][j-1]

    def sum_region(r1, c1, r2, c2):
        r1, c1, r2, c2 = r1+1, c1+1, r2+1, c2+1
        return dp[r2][c2] - dp[r2][c1-1] - dp[r1-1][c2] + dp[r1-1][c1-1] 
    
    return dp

def max_range_sum(arr, i=0, cumsum=0):
    # #find maximum sum that can be achieved by adding a range [i, j] of the array
    dp = [0 for i in range(len(arr))]
    dp[0] = max(0, arr[0])
    for i in range(1, len(arr)):
        dp[i] = max(dp[i-1] + arr[i], 0)
    return max(dp)

# print(arr)
# build(0, len(arr) - 1)
# print(tree)
# update(0, len(arr) - 1, 3, 10)
# print(tree)